File: 백준/Gold/2589. 보물섬/보물섬.py
from collections import deque

n, m = map(int, input().split())
board = [list(input()) for _ in range(n)]

maxi = 0

for y in range(n):
    for x in range(m):  # 여기를 수정: range(n) -> range(m)
        if board[y][x] == 'L':
            visited = [[0 for i in range(m)] for j in range(n)]
            dist = [[0 for i in range(m)] for j in range(n)]

            queue = deque()
            queue.append([y, x])
            visited[y][x] = 1
            while queue:
                ey, ex = queue.popleft()

                for dy, dx in [[0, 1], [1, 0], [-1, 0], [0, -1]]:
                    ny, nx = ey + dy, ex + dx
                    if 0 <= ny < n and 0 <= nx < m:
                        if board[ny][nx] == "L":
                            if visited[ny][nx] == 0:
                                visited[ny][nx] = 1
                                dist[ny][nx] = dist[ey][ex] + 1
                                maxi = max(maxi, dist[ny][nx])
                                queue.append([ny, nx])

print(maxi)

File: 백준/Gold/2589. 보물섬/보물섬.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static char[][] board;

    static int length = Integer.MIN_VALUE;

    static int[] dy = {-1, 0, 1, 0};
    static int[] dx = {0, -1, 0, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        board = new char[n][m];

        for(int i=0; i<n; i++) {
            String line = br.readLine();
            for(int j=0; j<m; j++) {
                board[i][j] = line.charAt(j);
            }
        }

        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                if (board[i][j] == 'L') {
                    bfs(i, j);
                }
            }
        }

        System.out.println(length - 1);
    }

    public static void bfs(int y, int x) {
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[] {y, x});
        int[][] visited = new int[n][m];
        visited[y][x] = 1;


        while(!queue.isEmpty()) {
            int[] curr = queue.poll();
            for(int i=0; i<4; i++) {
                int ny = curr[0] + dy[i];
                int nx = curr[1] + dx[i];

                if (ny >= 0 && ny < n && nx >= 0 && nx < m && board[ny][nx] == 'L' && visited[ny][nx] == 0) {
                    queue.offer(new int[] {ny, nx});
                    visited[ny][nx] = visited[curr[0]][curr[1]] + 1;
                    length = Math.max(length, visited[curr[0]][curr[1]] + 1);
                }
            }
        }

    }
}
